url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/arithmetic/ArithmeticParser.java

```diff
@@ -0,0 +1,276 @@
+	 */
+	protected Parser minusTerm() {
+		Sequence s = new Sequence();
+		s.add(new Symbol('-').discard());
+		s.add(term());
+		s.setAssembler(new MinusAssembler());
+		return s;
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *
+	 *    phrase = '(' expression ')' | Num;
+	 *
+	 * This parser adds an assembler to Num, that will 
+	 * replace the top token in the stack with the token's
+	 * Double value.
+	 */
+	protected Parser phrase() {
+		Alternation phrase = new Alternation("phrase");
+
+		Sequence s = new Sequence();
+		s.add(new Symbol('(').discard());
+		s.add(expression());
+		s.add(new Symbol(')').discard());
+		phrase.add(s);
+
+		phrase.add(new Num().setAssembler(new NumAssembler()));
+		return phrase;
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *    
+	 *     plusTerm = '+' term;
+	 *
+	 * This parser has an assembler that will pop two 
+	 * numbers from the stack and push their sum.
+	 */
+	protected Parser plusTerm() {
+		Sequence s = new Sequence();
+		s.add(new Symbol('+').discard());
+		s.add(term());
+		s.setAssembler(new PlusAssembler());
+		return s;
+	}
+
+	/**
+	 * Returns a parser that will recognize an arithmetic
+	 * expression.
+	 * 
+	 * @return   a parser that will recognize an 
+	 *           arithmetic expression
+	 */
+	public static Parser start() {
+		return new ArithmeticParser().expression();
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *
+	 *    term = factor (timesFactor | divideFactor)*;
+	 */
+	protected Parser term() {
+		Sequence s = new Sequence("term");
+		s.add(factor());
+
+		Alternation a = new Alternation();
+		a.add(timesFactor());
+		a.add(divideFactor());
+
+		s.add(new Repetition(a));
+		return s;
+	}
+
+	/*
+	 * Returns a parser that for the grammar rule:
+	 *    
+	 *     timesFactor = '*' factor;
+	 *
+	 * This parser has an assembler that will pop two 
+	 * numbers from the stack and push their product.
+	 */
+	protected Parser timesFactor() {
+		Sequence s = new Sequence();
+		s.add(new Symbol('*').discard());
+		s.add(factor());
+		s.setAssembler(new TimesAssembler());
+		return s;
+	}
+
+	/**
+	 * Return the value of an arithmetic expression given in a
+	 * string. This method is a façade, which provides an 
+	 * example of how to use the parser.
+	 *
+	 * @return the value of an arithmetic expression given in a
+	 *         string
+	 *
+	 * @param String the string to evaluate.
+	 *
+	 * @exception ArithmeticExpressionException if this 
+	 *            parser does not recognize the given string
+	 *            as a valid expression
+	 */
+	public static double value(String s) throws ArithmeticExpressionException {
+
+		TokenAssembly ta = new TokenAssembly(s);
+		Assembly a = start().completeMatch(ta);
+		if (a == null) {
+			throw new ArithmeticExpressionException("Improperly formed arithmetic expression");
+		}
+		Double d;
+		try {
+			d = (Double) a.pop();
+		} catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)